### PR TITLE
Ice anomaly spawns ice underneath it

### LINKED ISSUE #19713: Ice anomaly should make ice on the ground.
## Description

just do what the flesh anomaly does.

it would be very funny

### PR DESCRIPTION
**Pull request by YuriyKiss**
## About the PR

Close #19713
Ice anomaly now spawns ice tiles underneath just like flesh anomaly spawns flesh tiles.

## Why / Balance

I found that to be an interesting and easy feature to implement while I figure out how everything works in this repo.

As issue stated: "It would be very funny". Yeah, I mainly agree with that. This may be unbalanced tho.

## Technical details
Added new FrictionContactsSystem which works "kinda" similar to SlowContactsSystem; Friction values are being modified through MovementSpeedModifierSystem. From now on ice anomalies will spawn new ice_crust entity (similar to spiderweb entity) with that FrictionContacts component attached. Whenever an entity with MovementSpeedModifier component will come in contact with ice crust it's friction will be affected, those won't be changed until entity leaves ice_crust entities' colliders.

## Media
First media:
image

Final result:
https://github.com/space-wizards/space-station-14/assets/59531932/6aab9408-2762-4f97-ac01-9baa507bd9c7
- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

**Changelog**
- add: Ice anomaly spawns ice underneath it

### CODE REVIEW

**Comment by TheShuEd** [comment]
won't that cause problems? I don't think ice tiles can be removed by players

**Comment by UbaserB** [comment]
Current ice sprite looks bad, so i dont like it.

**Comment by YuriyKiss** [comment]
> I don't think ice tiles can be removed by players

Yes, you are right and that creates quite an issue:
image

And I suppose it won't be worth creating separate *proper* tile prototype for station with canCrowbar set to true and other properties set somewhat like in FloorFlesh. It is possible tho, I mean having two separate ice tiles for planet and station.

E.g. FloorFlesh prototype which is different in many ways I don't know yet:
image

**Comment by mirrorcult** [comment]
should make a new ice-type tile that does have plating underneath

**Comment by UbaserB** [comment]
> should make a new ice-type tile that does have plating underneath

thistbh

**Comment by YuriyKiss** [comment]
> should make a new ice-type tile that does have plating underneath

Should it drop ice tile item? Don't know whether this would be okay. In case we want ice to be breakable and drop nothing - can't do much here, engine's serializer prohibits that.

**Comment by TheShuEd** [comment]
Ideally, I think it should cover existing tiles with a thin layer of ice on top, much like spider webs do.  And this ice can be broken without destroying the tile underneath it

**Comment by metalgearsloth** [comment]
Making ice an entity is probably fine, I thought it would've been already?

**Comment by YuriyKiss** [comment]
Sooo, yeah, I've meddled around making ice an entity and here's what I've got:
https://github.com/space-wizards/space-station-14/assets/59531932/6e94b363-710a-4ba5-b1a3-7916e42f5d60

As you can see there are two problems here:
- sprites which I used suck. At least for the purpose we need 'em
- ice is not ice-y, I mean only tiles can have this friction effect, there is nothing similar for entities. 
For testing purposes I left component which cobweb uses.

Can someone give a hint what I can do to add friction? Mine hacky idea so far:
Modify tile under ice crust entity on spawn to have friction effect and restore original data whenever ice is destroyed.
Don't even know if that's possible, btw.

**Comment by YuriyKiss** [comment]
Not ready yet with some progress in, funny little friction lag-bug is there

**Comment by TheShuEd** [comment]
A seemingly simple idea turned out to be a much more complex task...

Oopsie

I'm glad there's progress ❤️

**Comment by YuriyKiss** [comment]
Yeah :) I've actually found another way to do that and implemented that my way which didn't need the use of RefreshMovementSpeedModifiersEvent and FrictionByContactComponent. Both solutions however seem to not work due to some black woodoo magic of physics code (sorry for quality, 10MB limitation bruh):

https://github.com/space-wizards/space-station-14/assets/59531932/13b440f1-53ca-48f9-9d3f-18a87389dfd3

Any ideas?

**Comment by TheShuEd** [comment]
Maybe prediction problem?

**Comment by TheShuEd** [comment]
maybe make the ice translucent?

**Comment by YuriyKiss** [comment]
I don't think that will help much. Most likely separate sprite is needed. I'll try to reach out to someone on Discord with that

**Comment by TheShuEd** [comment]
just
- type: Sprite
  color: "#ffffff44"

**Comment by forgotmyotheraccount** [comment]
How bad would it look if they shared puddle sprites?

**Comment by YuriyKiss** [comment]
mmm, okay, let me try that color modifier

**Comment by YuriyKiss** [comment]
It looks like that now:
https://github.com/space-wizards/space-station-14/assets/59531932/6aab9408-2762-4f97-ac01-9baa507bd9c7

**Comment by TheShuEd** [comment]
Perfect!

File: Content.Server/Anomaly/Effects/EntityAnomalySystem.cs [hunk 1]
```diff
@@ -21,27 +21,46 @@ public override void Initialize()
...
-        // A cluster of monsters
-        SpawnMonstersOnOpenTiles(component, xform, component.MaxSpawnAmount, component.SpawnRange, component.Spawns);
+        // A cluster of entities
+        SpawnEntitesOnOpenTiles(component, xform, component.MaxSpawnAmount, component.SpawnRange, component.Spawns);
         // And so much meat (for the meat anomaly at least)
-        SpawnMonstersOnOpenTiles(component, xform, component.MaxSpawnAmount, component.SpawnRange, component.SuperCriticalSpawns);
+        SpawnEntitesOnOpenTiles(component, xform, component.MaxSpawnAmount, component.SpawnRange, component.SuperCriticalSpawns);
+    }
+
+    private void OnSeverityChanged(EntityUid uid, EntitySpawnAnomalyComponent component, ref AnomalyStabilityChangedEvent args)
```
Comment (EmoGarbage404) [inline Content.Server/Anomaly/Effects/EntityAnomalySystem.cs:51]: fix the name of the function header to reflect the event

File: Content.Shared/Anomaly/Effects/Components/EntitySpawnAnomalyComponent.cs [hunk 2]
```diff
@@ -36,8 +33,21 @@ public sealed partial class EntitySpawnAnomalyComponent : Component
     public float SpawnRange = 5f;
 
     /// <summary>
-    /// The tile that is spawned by the anomaly's effect
+    /// Whether or not anomaly spawns entities on Pulse
     /// </summary>
-    [DataField("floorTileId", customTypeSerializer: typeof(PrototypeIdSerializer<ContentTileDefinition>)), ViewVariables(VVAccess.ReadWrite)]
-    public string FloorTileId = "FloorFlesh";
+    [DataField("spawnOnPulse"), ViewVariables(VVAccess.ReadWrite)]
```
Comment (EmoGarbage404) [inline Content.Shared/Anomaly/Effects/Components/EntitySpawnAnomalyComponent.cs:38]:
```suggestion
    [DataField, ViewVariables(VVAccess.ReadWrite)]
```

datafield names are inferred if left blank. juts remove it and it should work

File: Content.Shared/Anomaly/Effects/Components/EntitySpawnAnomalyComponent.cs [hunk 3]
```diff
@@ -36,8 +33,21 @@ public sealed partial class EntitySpawnAnomalyComponent : Component
...
+    public bool SpawnOnPulse = true;
+
+    /// <summary>
+    /// Whether or not anomaly spawns entities on SuperCritical
+    /// </summary>
+    [DataField("spawnOnSuperCritical"), ViewVariables(VVAccess.ReadWrite)]
```
Comment (EmoGarbage404) [inline Content.Shared/Anomaly/Effects/Components/EntitySpawnAnomalyComponent.cs:44]: here too

File: Content.Shared/Anomaly/Effects/Components/EntitySpawnAnomalyComponent.cs [hunk 4]
```diff
@@ -36,8 +33,21 @@ public sealed partial class EntitySpawnAnomalyComponent : Component
...
+    public bool SpawnOnSuperCritical = true;
+
+    /// <summary>
+    /// Whether or not anomaly spawns entities on StabilityChanged
+    /// The idea was to spawn entities either on Pulse/Supercritical OR StabilityChanged
+    /// </summary>
+    [DataField("spawnOnStabilityChanged"), ViewVariables(VVAccess.ReadWrite)]
```
Comment (EmoGarbage404) [inline Content.Shared/Anomaly/Effects/Components/EntitySpawnAnomalyComponent.cs:51]: and here

File: Resources/Textures/Objects/Misc/ice_crust.rsi/meta.json [hunk 5]
```diff
@@ -0,0 +1,14 @@
+{
+  "version": 1,
+  "license": "CC-BY-SA-3.0",
+  "copyright": "Taken from space-station-14+Resources+Textures+Tiles+Planet+Snow",
```
Comment (EmoGarbage404) [inline Resources/Textures/Objects/Misc/ice_crust.rsi/meta.json:4]: use the copyright listed in the attributions.yml file for the ice floor.

**Review by EmoGarbage404 — DISMISSED** [review]

File: Resources/Textures/Objects/Misc/ice_crust.rsi/meta.json (hunk 5 above)
Comment (YuriyKiss) [reply Resources/Textures/Objects/Misc/ice_crust.rsi/meta.json:4]: image
I would do that, but I've not found atributions.yml file in that folder and meta.json has no info :)

Commit where this file was added did not include any copyright claims as well:
https://github.com/space-wizards/space-station-14/pull/13487/files#diff-22a6f3bc9bafb71b461840aa1ebfb0f7f625ef81339732f58b7880111aaeb4d9

**Comment by Emisse** [comment]
looks good to merge when conflicts are cleaned up

**Review by Emisse — APPROVED** [review]

**Pull request merged**